src/main/java/org/jboss/logmanager/Bootstrap.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ *
+ * Copyright 2017 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jboss.logmanager;
+
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+import java.util.logging.Filter;
+import java.util.logging.Handler;
+
+import org.jboss.logmanager.handlers.FileHandler;
+
+/**
+ * A bootstrap helper that queues messages until the {@link #complete()} method is invoked.
+ *
+ * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
+ */
+class Bootstrap {
+
+    private final LoggerNode loggerNode;
+    private final Supplier<? extends Handler> handlerSupplier;
+    private final BootstrappedLogRecordPublisher publisher;
+    private final LogRecordPublisher handoffPublisher;
+    private final boolean isCalculateCaller;
+    private volatile BootstrapLogTask bootstrapLogTask;
+
+    /**
+     * Creates a new bootstrap.
+     *
+     * @param loggerNode    the logger node to bootstrap
+     * @param configuration the configuration
+     */
+    Bootstrap(final LoggerNode loggerNode, final BootstrapConfiguration configuration) {
+        this.loggerNode = loggerNode;
+        this.handlerSupplier = configuration.getHandler();
+        handoffPublisher = LogRecordPublisher.DEFAULT;
+        publisher = new BootstrappedLogRecordPublisher(configuration.getQueueSize());
+        isCalculateCaller = configuration.isCalculateCaller();
+        loggerNode.setLogRecorderPublisher(publisher);
+        loggerNode.setLevel(configuration.getLevel());
+        bootstrapLogTask = registerShutdownHook();
+    }
+
+    /**
+     * Completes the bootstrapping by pushing queued messages to the {@linkplain LogRecordPublisher#DEFAULT default}
+     * publisher then replaces the publisher on all child logger nodes.
+     */
+    void complete() {
+        unregisterShutdownHook();
+        final Object lock = loggerNode.getContext().treeLock;
+        synchronized (publisher) {
```

Comment:
If the publisher field lives on the log context, the use of the tree lock here can go away (I think).